### PR TITLE
Fix minimal accessibility failures

### PR DESCRIPTION
**Pull request by Alchav**
The problem with minimal accessibility has been that items from a world that has items/location accessibility can place into a world with minimal accessibility, and then afterwards, items that the minimal accessibility world would need to reach them can be locked somewhere considered not accessible (as long as the multiworld is still beatable).

This PR adds a function that runs after priority location fill and the main prog item fill which checks all unreachable locations in all worlds with minimal accessibility, pulls out any progression items for players that do NOT have minimal accessibility, and runs them back through fill_restrictive again in order to place them again.

### CODE REVIEW

**Comment by Alchav** [comment]
Using `--seed 2` the following yaml failed generation on main during accessibility check, but succeeded on this branch:

```
name: a
game: A Link to the Past
A Link to the Past:
  accessibility: minimal
  goals: local_triforce_hunt
  local_items: [Hookshot, Progressive Bow, Progressive Bow (Alt), Hammer, Moon Pearl, Progressive Glove, Progressive Sword, Magic Mirror, Book of Mudora, Lamp, Fire Rod, Pegasus Boots, Flippers, Cane of Somaria]
---
name: b
game: A Link to the Past
A Link to the Past:
  accessibility: locations
  smallkey_shuffle: different_world
  bigkey_shuffle: different_world
  non_local_items: [Hookshot, Progressive Bow, Progressive Bow (Alt), Hammer, Moon Pearl, Progressive Glove, Progressive Sword, Magic Mirror, Book of Mudora, Lamp, Fire Rod, Pegasus Boots, Flippers, Cane of Somaria]
```

**Comment by Berserker66** [comment]
@espeon65536 I think this is something you were looking into?

File: Fill.py [hunk 1]
```diff
@@ -128,6 +129,34 @@ def fill_restrictive(world: MultiWorld, base_state: CollectionState, locations:
     itempool.extend(unplaced_items)
 
 
+def accessibility_corrections(world: MultiWorld, state: CollectionState, locations, pool=[]):
+    maximum_exploration_state = sweep_from_pool(state, pool)
+    minimal_players = {player for player in world.player_ids if world.accessibility[player] == "minimal"}
+    unreachable_locations = [location for location in world.get_locations() if location.player in minimal_players and
+                             not location.can_reach(maximum_exploration_state)]
+    for location in unreachable_locations:
+        if (location.item is not None and location.item.classification == ItemClassification.progression and
```
Comment (Berserker66) [inline Fill.py:138]: Should this not be `ItemClassification.progression in location.item.classification` ?

File: Fill.py [hunk 2]
```diff
@@ -128,6 +129,34 @@ def fill_restrictive(world: MultiWorld, base_state: CollectionState, locations:
...
+
+    if pool:
+        fill_restrictive(world, state, locations, pool)
+
+
+def inaccessible_location_rules(world: MultiWorld, state: CollectionState, locations):
+    maximum_exploration_state = sweep_from_pool(state, [])
+    unreachable_locations = [location for location in locations if not location.can_reach(maximum_exploration_state)]
+    for location in unreachable_locations:
+        add_item_rule(location, lambda item: not ((item.useful or item.advancement) and
```
Comment (Berserker66) [inline Fill.py:156]: `item.useful or item.advancement` can be expressed as the respective bit flags for some efficiency improvement.

File: Fill.py (hunk 1 above)
Comment (Alchav) [reply Fill.py]: Changed it to location.item.advancement

File: Fill.py (hunk 2 above)
Comment (Alchav) [reply Fill.py]: changed to `item.classification & 0b0011`

**Comment by Jarno458** [comment]
We should probely link this #434

**Comment by Alchav** [comment]
> We should probely link this #434

This will not fix that issue. I'm assuming what is causing that issue is that an item is being placed in an unreachable locations, locking out whatever areas of the game that item would be needed to reach, and reducing the number of reachable locations to fewer than the number of advancement items that need to be placed. I have no solution for that.

**Comment by espeon65536** [comment]
Finally got around to looking at this. It looks good to me; although I'm not completely certain it always fixes seeds, the only counterexamples I can generate are pathological cases where there are more priority locations than progression items. I also think this may fix #434 since those generations fail during the location accessibility check, which this should prevent by searching for such failures and repairing them. Either way, it's an improvement over what we have right now, so it seems fine to merge.

**Comment by Berserker66** [comment]
image

**Comment by Alchav** [comment]
> image

fixed

**Review by Berserker66 — APPROVED** [review]

**Pull request merged**